docs/async/async-csharp-vb.rst

```diff
@@ -0,0 +1,213 @@
+
+0-1--------------------------------------------------------------------------------------------------------------------------------------------------2-3
+
+* Time spent from points ``0`` to ``1`` is everything up until an async method yield control to its caller.
+
+* Time spent from points ``1`` to ``2`` is the time spend on I/O.
+
+* Finally, time spent from ``2`` to ``3`` is passing control back to the async method, at which point it is executing again.
+
+Tasks are also used outside of the async programming model.  They are the foundation of the Task Parallel Library, which supports the parallelization of CPU-bound work via `Data Parallelism <https://msdn.microsoft.com/en-us/library/dd537608(v=vs.110).aspx>`_ and `Task Parallelism <https://msdn.microsoft.com/en-us/library/dd537609(v=vs.110).aspx>`_.
+
+Important Info and Advice
+-------------------------
+
+Although async programming is relatively straightforward, there are some details to keep in mind which can prevent unexpected behavior.
+
+* **You should add "Async" as the suffix of every async method name you write.**
+
+This is the convention used in .NET to more-easily differentiate synchronous and asynchronous methods.  Note that certain methods which aren't explicitly called by your code (such as event handlers or web controller methods) don't necessarily apply.  Because these are not explicitly called by your code, being explicit about their naming isn't as important.
+
+* ``async void`` **should only be used for event handlers.**
+
+Why?  It's the only way to allow asynchronous event handlers work because events do not have return types (thus cannot make use of ``Task`` and ``Task<T>``).  Any other use of ``async void``does not follow the Task-based model and will bring about some issues, such as:
+
+    (a) Exceptions thrown in an ``async void`` method can't be caught outside of that method.
+
+    (b) ``async void`` methods are very difficult to test.
+
+    (c) ``async void`` methods can cause bad side effects if the caller isn't expecting them to be async.
+
+* **Tread carefully when using async lambdas in LINQ expressions**
+
+Lambda expressions in LINQ use deferred execution, meaning code could end up executing at a time when you're not expecting it to.  The introduction of blocking tasks into this can easily result in a deadlock if not written correctly.  The nesting of asynchronous code like this can also make it more difficult to reason about the execution of the code.  Async and LINQ are powerful, but if they are abused they can make things difficult.  Clarity is always better than cleverness.
+
+* **Write code that awaits Tasks in a non-blocking manner**
+
+Blocking the current thread as a means to wait for a Task to complete can result in deadlocks, blocked context threads, and significantly more complex error-handling.  The following table should provide some guidance in how to deal with waiting for Tasks in a non-blocking way:
```

Comment:
Do you mean "and required significantly more complex ..."

You can say: "The following table provides guidance on how to deal with ..." We know it provides the guidance. It's not a question of "should".